src/service-worker.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import {writeFile} from 'fs';
+import * as path from 'path';
+import * as logging from 'plylog';
+import {PolymerProject} from './polymer-project';
+import {DepsIndex} from './analyzer';
+const swPrecache = require('sw-precache');
+
+// non-ES compatible modules
+let logger = logging.getLogger('polymer-build.service-worker');
+
+export interface SWConfig {
+  cacheId?: string;
+  directoryIndex?: string;
+  dynamicUrlToDependencies?: {
+    [property: string]: string[]
+  };
+  handleFetch?: boolean;
+  ignoreUrlParametersMatching?: RegExp[];
+  importScripts?: string[];
+  logger?: Function;
+  maximumFileSizeToCacheInBytes?: number;
+  navigateFallback?: string;
+  navigateFallbackWhitelist?: RegExp[];
+  replacePrefix?: string;
+  runtimeCaching?: {
+    urlPattern: RegExp;
+    handler: string;
+    options?: {
+      cache: {
+        maxEntries: number;
+        name: string;
+      };
+    };
+  }[];
+  staticFileGlobs?: string[];
+  stripPrefix?: string;
+  templateFilePath?: string;
+  verbose?: boolean;
+}
+
+export interface AddServiceWorkerOptions {
+  project: PolymerProject;
+  buildRoot: string;
+  bundled?: boolean;
```

Comment:
+1 point for the build stream as its own custom type :)

We could split the two methods and add `addBundledServiceWorker()`. `generateServiceWorker()` could remain as-is so that we don't need to split that as well.

We could also make `getPrecachedAssets()`/`getBundledPrecachedAssets()` public and have the gulpfile author properly call those themselves, and then send the result as an option to `generateServiceWorker()` instead.

I don't love it either but it's a small nit for me vs. our other options. Let me know if you think this is a big enough problem that we should address it now.